client/schema.go

```diff
@@ -0,0 +1,405 @@
+	enumID, ok = cf.enumMappings[tableName][columnID][enumCase]
+	cf.RUnlock()
+	if !ok {
+		cf.metricScope.Tagged(
+			map[string]string{
+				"TableName": tableName,
+				"ColumnID":  strconv.Itoa(columnID),
+			},
+		).Counter("new_enum_case_rows_ignored").Inc(int64(1))
+		if defaultValue, ok := cf.enumDefaultValueMappings[tableName][columnID]; ok {
+			return defaultValue, nil
+		}
+		return -1, nil
+	}
+	return enumID, nil
+}
+
+// FetchAllSchema fetch all schemas
+func (cf *CachedSchemaHandler) FetchAllSchema() error {
+	tables, err := cf.schemaFetcher.FetchAllSchemas()
+	if err != nil {
+		return err
+	}
+
+	for _, table := range tables {
+		cf.setTable(table)
+		err := cf.fetchAndSetEnumCases(table)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// FetchSchema fetchs the schema of given table name
+func (cf *CachedSchemaHandler) FetchSchema(tableName string) (*TableSchema, error) {
+	cf.RLock()
+	schema, exist := cf.schemas[tableName]
+	cf.RUnlock()
+	if exist {
+		return schema, nil
+	}
+	table, err := cf.schemaFetcher.FetchSchema(tableName)
+	if err != nil {
+		return nil, err
+	}
+	schema = cf.setTable(table)
+	err = cf.fetchAndSetEnumCases(table)
+	return schema, err
+}
+
+// PrepareEnumCases prepares enum cases
+func (cf *CachedSchemaHandler) PrepareEnumCases(tableName, columnName string, enumCases []string) error {
+	newEnumCases := make([]string, 0, len(enumCases))
+	cf.RLock()
+	schema, exist := cf.schemas[tableName]
+	if !exist {
+		cf.RUnlock()
+		return nil
+	}
+	columnID, exist := schema.ColumnDict[columnName]
+	if !exist {
+		cf.RUnlock()
+		return nil
+	}
+
+	caseInsensitive := schema.Table.Columns[columnID].CaseInsensitive
+	disableAutoExpand := schema.Table.Columns[columnID].DisableAutoExpand
+	for _, enumCase := range enumCases {
+		if _, exist := cf.enumMappings[tableName][columnID][enumCase]; !exist {
+			newEnumCases = append(newEnumCases, enumCase)
+		}
+	}
+	cf.RUnlock()
+
+	if disableAutoExpand {
+		// It's recommended to set up elk or sentry logging to catch this error.
+		cf.logger.With(
+			"TableName", tableName,
+			"ColumnName", columnName,
+			"ColumnID", columnID,
+			"newEnumCasesSet", newEnumCases,
+			"caseInsensitive", caseInsensitive,
+		).Error("Finding new enum cases during ingestion but enum auto expansion is disabled")
+		cf.metricScope.Tagged(
+			map[string]string{
+				"TableName": tableName,
+				"ColumnID":  strconv.Itoa(columnID),
+			},
+		).Counter("new_enum_cases_ignored").Inc(int64(len(newEnumCases)))
+		return nil
+	}
+
+	enumIDs, err := cf.schemaFetcher.ExtendEnumCases(tableName, columnName, newEnumCases)
+	if err != nil {
+		return err
+	}
+
+	cf.Lock()
+	for index, enumCase := range newEnumCases {
+		if caseInsensitive {
+			enumCase = strings.ToLower(enumCase)
+		}
+		cf.enumMappings[tableName][columnID][enumCase] = enumIDs[index]
+	}
+	cf.Unlock()
+	return nil
+}
+
+func (cf *CachedSchemaHandler) fetchAndSetEnumCases(table *metaCom.Table) error {
```

Comment:
also, possible to combine to fetch all column enums in one http call?

Reply:
maybe later? currently we do not support combined call in the api. plus this is not really a performance bottleneck.